### PR TITLE
feat(entity): publish events of upsert batch of entities #255

### PR DESCRIPTION
**Pull request by HoucemKacem**

### CODE REVIEW

File: entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt [hunk 1]
```diff
@@ -150,28 +150,29 @@ class EntityOperationService(
...
         neo4jRepository.deleteEntityAttributes(entity.id)
         val (_, notUpdated) = entityService.appendEntityAttributes(entity.id, entity.attributes, false)
         if (notUpdated.isEmpty()) {
-            return entity.id.right()
+            return BatchEntityReplaceSuccess(
+                entity.id
+            )
         } else {
             throw BadRequestDataException(
-                ArrayList(notUpdated.map { it.attributeName + " : " + it.reason }).joinToString()
```
Comment (HoucemKacem) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt:160]: Is it supposed to be returned in expanded format ?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt [hunk 2]
```diff
@@ -150,28 +151,30 @@ class EntityOperationService(
...
         neo4jRepository.deleteEntityAttributes(entity.id)
         val (_, notUpdated) = entityService.appendEntityAttributes(entity.id, entity.attributes, false)
-        if (notUpdated.isEmpty()) {
-            return entity.id.right()
-        } else {
+        if (notUpdated.isEmpty())
+            return BatchEntitySuccess(entity.id).right()
+        else
             throw BadRequestDataException(
                 ArrayList(notUpdated.map { it.attributeName + " : " + it.reason }).joinToString()
```
Comment (HoucemKacem) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt:161]: Is it supposed to be returned in expanded format ?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt [hunk 3]
```diff
@@ -145,11 +152,46 @@ class EntityOperationHandler(
         return ResponseEntity.status(HttpStatus.OK).body(batchOperationResult)
     }
 
-    private fun extractAndParseBatchOfEntities(payload: String): Pair<List<Map<String, Any>>, List<NgsiLdEntity>> {
-        val extractedEntities = JsonUtils.parseListOfEntities(payload)
-        return Pair(
-            extractedEntities,
-            JsonLdUtils.expandJsonLdEntities(extractedEntities).map { it.toNgsiLdEntity() }
-        )
+    private fun extractAndParseBatchOfEntities(payload: String) =
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt:155]: When it is not obvious (and it is not here), add explicitely the return type in the function signature

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt [hunk 4]
```diff
@@ -50,8 +47,8 @@ class EntityOperationHandler(
             }
         )
 
-        authorizationService.createAdminLinks(batchOperationResult.success, userId)
-        ngsiLdEntities.filter { it.id in batchOperationResult.success }
+        authorizationService.createAdminLinks(batchOperationResult.success.map { it.entityId }, userId)
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt:50]: you're doing this `batchOperationResult.success.map { it.entityId }` in a lot of places, I think you could make it a function of the `BatchOperationResult` data class.

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt [hunk 5]
```diff
@@ -145,11 +152,46 @@ class EntityOperationHandler(
...
+            val jsonLdEntity = jsonLdEntities.find { jsonLdEntity -> jsonLdEntity.id.toUri() == it.entityId }!!
+            entityEventService.publishAppendEntityAttributesEvents(
+                it.entityId,
+                jsonLdEntity.properties,
+                it.updateResult!!,
+                entityOperationService.getFullEntityById(it.entityId, true)!!,
+                jsonLdEntity.contexts
+            )
+        }
+        updateBatchOperationResult.errors.filter { it.updateResult!!.updated.isNotEmpty() }.forEach {
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt:186]: successes inside errors, that looks weird

Comment (bobeal) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt:186]: the granularity of a batch operation should be the entity (or so I think). whether it has been fully updated, whether it has not been updated at all. I don't think a mix of success and errors is compliant.

Comment (bobeal) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityOperationHandler.kt:186]: so probably `EntityOperationService.updateEntity(...)` should be transactional (as is `replaceEntity`)?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/APIResponses.kt [hunk 6]
```diff
@@ -1,19 +1,33 @@
...
+    @JsonIgnore
+    val updateResult: UpdateResult? = null
+)
+
 data class BatchEntityError(
     val entityId: URI,
-    val error: MutableList<String>
+    val error: MutableList<String>,
+    @JsonIgnore
+    val updateResult: UpdateResult? = null
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/APIResponses.kt:32]: do you still need this attribute in `BatchEntityError`?

Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/APIResponses.kt:32]: indeed, i'll remove it.

**Review by bobeal — APPROVED** [review]

File: entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt (hunk 2 above)
Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt:161]: The specification mention that attributes should be represented by their names.
So I'm not sure about the format.

Comment (bobeal) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt:161]: not part of the PR, leave them as they are currently.

**Pull request merged**